tools/Dashboard/DevHome.Dashboard/Helpers/WidgetIconCache.cs

```diff
@@ -0,0 +1,131 @@
+        {
+            await AddIconsToCache(widgetDef);
+        }
+    }
+
+    /// <summary>
+    /// Caches two icons for each widget, one for light theme and one for dark theme.
+    /// </summary>
+    public async Task AddIconsToCache(WidgetDefinition widgetDef)
+    {
+        // Only cache icons for providers that we're including.
+        if (WidgetHelpers.IsIncludedWidgetProvider(widgetDef.ProviderDefinition))
+        {
+            var widgetDefId = widgetDef.Id;
+            try
+            {
+                Log.Logger()?.ReportDebug("WidgetIconCache", $"Cache widget icons for {widgetDefId}");
+                var itemLightImage = await WidgetIconToBitmapImage(widgetDef.GetThemeResource(WidgetTheme.Light).Icon);
+                var itemDarkImage = await WidgetIconToBitmapImage(widgetDef.GetThemeResource(WidgetTheme.Dark).Icon);
+
+                // There is a widget bug where Definition update events are being raised as added events.
+                // If we already have an icon for this key, just remove and add again in case the icons changed.
+                if (_widgetLightIconCache.ContainsKey(widgetDefId))
+                {
+                    _widgetLightIconCache.Remove(widgetDefId);
+                }
+
+                if (_widgetDarkIconCache.ContainsKey(widgetDefId))
+                {
+                    _widgetDarkIconCache.Remove(widgetDefId);
+                }
+
+                _widgetLightIconCache.Add(widgetDefId, itemLightImage);
+                _widgetDarkIconCache.Add(widgetDefId, itemDarkImage);
+            }
+            catch (Exception ex)
+            {
+                Log.Logger()?.ReportError("WidgetIconCache", $"Exception in AddIconsToCache:", ex);
+                _widgetLightIconCache.Add(widgetDefId, null);
+                _widgetDarkIconCache.Add(widgetDefId, null);
+            }
+        }
+    }
+
+    public void RemoveIconsFromCache(string definitionId)
+    {
+        _widgetLightIconCache.Remove(definitionId);
+        _widgetDarkIconCache.Remove(definitionId);
+    }
+
+    public static BitmapImage GetWidgetIconForTheme(WidgetDefinition widgetDefinition, ElementTheme theme)
+    {
+        BitmapImage image;
+        if (theme == ElementTheme.Light)
+        {
+            _widgetLightIconCache.TryGetValue(widgetDefinition.Id, out image);
+        }
+        else
+        {
+            _widgetDarkIconCache.TryGetValue(widgetDefinition.Id, out image);
+        }
+
+        return image;
+    }
+
+    public static Brush GetBrushForWidgetIcon(WidgetDefinition widgetDefinition, ElementTheme theme)
+    {
+        var image = GetWidgetIconForTheme(widgetDefinition, theme);
+
+        var brush = new ImageBrush
+        {
+            ImageSource = image,
+        };
+        return brush;
+    }
+
+    private static async Task<BitmapImage> WidgetIconToBitmapImage(IRandomAccessStreamReference iconStreamRef)

```

Comment:
NIT: If this method is the only one that uses `_dispatcherQueue` maybe it can be an input parameter?

Reply:
The method that needs it gets called by other things but I still like just passing it into the methods so I made the change.